src/common.cpp

```diff
@@ -1888,39 +1886,83 @@ std::string get_executable_path(const char *argv0) {
-                                ok = true;
-                                break;
-                            }
-                        }
+    char errbuf[_POSIX2_LINE_MAX];
+    const char *pwd = nullptr, *cwd = nullptr, *penv = nullptr;
+    if (!std::string(argv0).empty()) {
+        if (argv0[0] == '/') {
+            path = argv0;
+            goto finish1;
+        } else if (std::string(argv0).find('/') == std::string::npos) {
+            penv = getenv("PATH");
+            if (penv && *penv) {
+                wcstring wenv = str2wcstring(penv);
+                wcstring_list_t env = split_string(wenv, L':');
+                for (std::size_t i = 0; i < env.size(); i++) {
+                    path = wcs2string(env[i].c_str(), env[i].length()) + "/" + argv0;
+                    if (!stat(path.c_str(), &st) && (st.st_mode & S_IXUSR) && (st.st_mode & S_IFREG)) {
+                        goto finish2;
+                    }
+                    path.clear();
+                }
+            }
+        }
+        pwd = getenv("PWD");
+        if (pwd && *pwd) {
+            path = std::string(pwd) + "/" + std::string(argv0);
+            if (!stat(path.c_str(), &st) && (st.st_mode & S_IXUSR) && (st.st_mode & S_IFREG)) {
+                goto finish2;
+            } else {
+                goto fallback;
+            }
+        } else {
+            fallback:
+            cwd = getcwd(exe, buff_size);
+            if (cwd && *cwd) {
+                path = std::string(cwd) + "/" + std::string(argv0);
+            }
+        }
+        finish1:
+        if (!path.empty()) {
+            if (!stat(path.c_str(), &st) && (st.st_mode & S_IXUSR) && (st.st_mode & S_IFREG)) {
+                finish2:
+                if (realpath(path.c_str(), exe)) {
+                    path = exe;
+                    goto finish3;
+                }
+            }
+        }
+    }
+    finish3:
+    if (path.empty()) goto finish4;
+    kd = kvm_openfiles(nullptr, nullptr, nullptr, KVM_NO_FILES, errbuf); 
+    if (!kd) { path.clear(); goto finish4; }
+    if ((kif = kvm_getfiles(kd, KERN_FILE_BYPID, getpid(), sizeof(struct kinfo_file), &cntp))) {
+        for (int i = 0; i < cntp; i++) {
+            if (kif[i].fd_fd == KERN_FILE_TEXT) {
+                if (!stat(path.c_str(), &st)) {
+                    if (st.st_dev == (dev_t)kif[i].va_fsid || st.st_ino == (ino_t)kif[i].va_fileid) {
+                        ok = true;
+                        break;
                     }
                 }
             }
-            kvm_close(kd);
         }
     }
-    narrow[buff_size] = '\0';
-    strcpy(buff, narrow.substr(0, buff_size).c_str());
-    return ((ok) ? std::string(buff)  : "fish");
+    if (!ok) { path.clear(); }
+    kvm_close(kd);
+    finish4:
+    path.resize(buff_size, '\0');
+    path[buff_size] = '\0';
+    strcpy(buff, path.c_str());
+    static std::string result = buff;
```

Comment:
Defining a static variable this late means you do all the computation anyway on the second go and then throw it all away because the assignment never happens.

Reply:
It's *fine* to use a static variable.

But what you would do is:

```fish
static std::string ret;

if (!ret.empty()) return ret;

// And now the rest of the function
```

That way you skip *all of the other code*.

Reply:
Like you add the static variable at the top of the function, before everything else.

If it has something in it you're in the second run and can just return.

Reply:
I added a static bool so that it only calculates the path once for each run of fish. Not what you asked, let me know if that's ok, it sould be more efficient than runnting multiple times that code logic.

Reply:
Ah. I didn't know static worked like that. I'm going to change it to your suggestion.